api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -623,9 +647,10 @@ private List<NestedField> lazyFieldList() {
 
     private Map<String, NestedField> lazyFieldsByName() {
       if (fieldsByName == null) {
+        Map<String, Integer> nameToId = TypeUtil.indexByName(this);
```

Comment:
While I believe this is the right way to go, it changes current API behavior. E.g. `field(String name)` now not only returns top level fields, but also nested fields. Maybe @szehon-ho @rdblue can help verify if that's ok?